thirdparty/github.go

```diff
@@ -81,6 +85,47 @@ const (
 	githubPRUnstable = "unstable"
 )
 
+var cachingTransport http.RoundTripper
+
+type cacheControlTransport struct {
+	base http.RoundTripper
+}
+
+// RoundTrip sets the [max-age] Cache-Control directive before passing the request off
+// to the base transport.
+//
+// [max-age]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control#max-age_2
+func (t *cacheControlTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	req.Header.Set("Cache-Control", "max-age=0")
+	return t.base.RoundTrip(req)
+}
+
+func init() {
+	// Base transport sends a request over the wire.
+	baseTransport := utility.DefaultTransport()
+
+	// Wrap in a transport that creates a new span for each request that goes over the wire.
+	otelTransport := otelhttp.NewTransport(baseTransport)
+
+	// Wrap in a transport that caches responses to reduce the number of calls that count against our rate limit.
+	memoryCacheTransport := httpcache.NewTransport(&cache.DBCache{})
+	memoryCacheTransport.Transport = otelTransport
+
+	// Wrap in a transport that overrides the cache-control header so we don't rely on max-age
+	// from GitHub's response to not ask if there's been a change.
```

Comment:
"so we don't rely on a response to not ask" -- this sentence kind of confuses me 😅  why don't we want to rely on max age? why can't we set cachingTransport here directly as opposed to having this extra variable maxAgeTransport?

Reply:
For some routes GitHub sets a max-age in the response to 60 seconds. This would mean that if we request the same thing from GitHub within 60 seconds the cache layer wouldn't even make a conditional request and instead just pull it out of the cache. This seems like a bad idea for our use case where things can change quickly. 

The idea of this extra transport is to override the max-age from the response by telling the cache that we want a max-age of 0. If the cache layer gets a request that sets max-age to something different than the response it will go with the request and make the conditional request according to the request.

Reply:
Thanks for the info! maybe we could reword to "so we don't use Github's max-age, which prevents us from asking if there's been a change if we requested recently" or something like that to remove the double negative?

Reply:
done